pkg/services/login/authinfoservice/service.go

```diff
@@ -0,0 +1,211 @@
+			if query.UserId != 0 && query.UserId != authQuery.Result.UserId {
+				err = s.Bus.Dispatch(&models.DeleteAuthInfoCommand{
+					UserAuth: authQuery.Result,
+				})
+				if err != nil {
+					s.logger.Error("Error removing user_auth entry", "error", err)
+				}
+
+				return false, nil, nil, models.ErrUserNotFound
+			} else {
+				has, user, err := s.getUserById(authQuery.Result.UserId)
+				if err != nil {
+					return false, nil, nil, err
+				}
+
+				if !has {
+					// if the user has been deleted then remove the entry
+					err = bus.Dispatch(&models.DeleteAuthInfoCommand{
+						UserAuth: authQuery.Result,
+					})
+					if err != nil {
+						s.logger.Error("Error removing user_auth entry", "error", err)
+					}
+
+					return false, nil, nil, models.ErrUserNotFound
+				}
+
+				return true, user, authQuery.Result, nil
+			}
+		}
+	}
+
+	return false, nil, nil, models.ErrUserNotFound
+}
+
+func (s *Implementation) LookupByOneOf(userId int64, email string, login string) (bool, *models.User, error) {
+	foundUser := false
+	var user *models.User
+	var err error
+
+	// If not found, try to find the user by id
+	if !foundUser && userId != 0 {
+		foundUser, user, err = s.getUserById(userId)
+		if err != nil {
+			return false, nil, err
+		}
+	}
+
+	// If not found, try to find the user by email address
+	if !foundUser && email != "" {
+		user = &models.User{Email: email}
+		foundUser, err = s.getUser(user)
+		if err != nil {
+			return false, nil, err
+		}
+	}
+
+	// If not found, try to find the user by login
+	if !foundUser && login != "" {
+		user = &models.User{Login: login}
+		foundUser, err = s.getUser(user)
+		if err != nil {
+			return false, nil, err
+		}
+	}
+
+	return foundUser, user, err
+}
+
+func (s *Implementation) GenericOAuthLookup(authModule string, authId string, userID int64) (*models.UserAuth, error) {
+	if authModule == genericOAuthModule && userID != 0 {
+		authQuery := &models.GetAuthInfoQuery{}
+		authQuery.AuthModule = authModule
+		authQuery.AuthId = authId
+		authQuery.UserId = userID
+		err := bus.Dispatch(authQuery)
+		if err != nil {
+			return nil, err
+		}
+
+		return authQuery.Result, nil
+	}
+	return nil, nil
+}
+
+func (s *Implementation) LookupAndUpdate(query *models.GetUserByAuthInfoQuery) (*models.User, error) {
+	// 1. LookupAndFix = auth info, user, error
+	// TODO: Not a big fan of the fact that we are deleting auth info here, might want to move that
+	foundUser, user, authInfo, err := s.LookupAndFix(query)
+	if err != nil && !errors.Is(err, models.ErrUserNotFound) {
+		return nil, err
+	}
+
+	// 2. FindByUserDetails
+	if !foundUser {
+		foundUser, user, err = s.LookupByOneOf(query.UserId, query.Email, query.Login)
+	}
+
+	// No user found
+	if !foundUser {
+		return nil, models.ErrUserNotFound
+	}
+
+	if allow, err := s.UserProtectionService.AllowUserMapping(user, query.AuthModule); err != nil {
+		return nil, err
+	} else if !allow {
+		return nil, models.ErrProtectedUser
+	}
```

Comment:
Could make sense to simplify the business logic and the signature of the `UserProtectionService` to make it return directly the error? (in case the mapping is not allowed it would return `models.ErrProtectedUser` instead of `nil` plus boolean) 🤔

Reply:
I agree, the bool just generates extra for no reason.